tiledb/bioimg/converters/scale.py

```diff
@@ -12,261 +12,102 @@
+
+        scaled_tile.append(slice(start, stop))
+
+    output[tile] = resize(
+        base[tuple(scaled_tile)],
+        slice_shape,
+        preserve_range=True,
+        order=order,
+        anti_aliasing=True,
+    )
 
 
 class Scaler(object):
     def __init__(
-        self, base: tiledb.array.DenseArray, axes: Axes, scale_factor: List[int]
+        self,
+        base: tiledb.array,
+        base_axes: Axes,
+        scale_factors: Sequence[float],
+        scale_axes: str,
+        mode: ScalerMode,
+        order: int,
     ):
-        self._scale_factor = scale_factor
-        self._scale_factor_progressive = [float(scale_factor[0])]
-
-        for index in range(1, len(scale_factor)):
-            self._scale_factor_progressive.append(
-                scale_factor[index] / float(scale_factor[index - 1])
-            )
-
-        self._initial_shape = base.shape
+        self._mode = mode
+        self._order = order
 
+        self._scale_factors = []
+        self._scale_factors_progressive = []
         self._resolutions = []
 
-        self._tiles = []
-        self._x_index = axes.dims.find("X")
-        self._y_index = axes.dims.find("Y")
-
-        for level in range(len(scale_factor)):
-            ratio = scale_factor[level]
-
-            shape = list(self._initial_shape)
-
-            shape[self._x_index] = int(
-                round(self._initial_shape[self._x_index] / float(ratio))
+        previous_scale_factor = 1.0
+
+        for factor in scale_factors:
+            level_factors = [
+                factor if axis in scale_axes else 1 for axis in base_axes.dims
+            ]
+            level_factors_progressive = [
+                factor / previous_scale_factor if axis in scale_axes else 1
+                for axis in base_axes.dims
+            ]
+            resolution = tuple(
+                round(base.shape[i] / level_factors[i]) for i in range(len(base.shape))
             )
-            shape[self._y_index] = int(
-                round(self._initial_shape[self._y_index] / float(ratio))
-            )
-
-            tiles = []
-
-            for tile in iter_tiles(base.domain):
-                scaled_tile = list(tile)
 
-                scaled_x_start = tile[self._x_index].start // ratio
-                scaled_x_end = tile[self._x_index].stop // ratio
-                scaled_tile[self._x_index] = slice(scaled_x_start, scaled_x_end)
+            previous_scale_factor = factor
 
-                scaled_y_start = tile[self._y_index].start // ratio
-                scaled_y_end = tile[self._y_index].stop // ratio
-                scaled_tile[self._y_index] = slice(scaled_y_start, scaled_y_end)
-
-                tiles.append(tuple(tile))
-
-            self._resolutions.append(tuple(shape))
-            self._tiles.append(tiles)
+            self._scale_factors.append(level_factors)
+            self._scale_factors_progressive.append(level_factors_progressive)
```

Comment:
You don't need to compute both `_scale_factors` and `_scale_factors_progressive`; you already know the mode so you can tell which are needed (or neither for non-chunked mode).